blog/_posts/2019-01-18-fluxdiffeq.md

```diff
@@ -0,0 +1,657 @@
+[CASADI](https://web.casadi.org/), and
+[CVODES](https://computation.llnl.gov/projects/sundials/cvodes)
+have been available with this adjoint method for a long time (CVODES came out
+in 2005!). [DifferentialEquations.jl has sensitivity analysis implemented too](http://docs.juliadiffeq.org/latest/analysis/sensitivity.html)
+
+The problem with adjoint sensitivity analysis methods is that they require
+multiple forward solutions of the ODE. As you would expect, this is very costly.
+Methods like the checkpointing scheme in CVODES reduce the cost by saving closer
+time points to make the forward solutions shorter at the cost of using more
+memory. The method in the neural ordinary differential equations paper tries to
+eliminate the need for these forward solutions by doing a backwards solution
+of the ODE itself along with the adjoints. The issue with this is that this
+method implicitly makes the assumption that the ODE integrator is
+[reversible](https://www.physics.drexel.edu/~valliere/PHYS305/Diff_Eq_Integrators/time_reversal/).
+Sadly, there are no reversible adaptive integrators for first-order ODEs, so
+with no ODE solver method is this guaranteed to work. For example, here's a quick
+equation where a backwards solution to the ODE using the Adams method from the
+paper has >1700% error in its final point, even with solver tolerances of 1e-12:
+
+```julia
+using Sundials, DiffEqBase
+function lorenz(du,u,p,t)
+ du[1] = 10.0*(u[2]-u[1])
+ du[2] = u[1]*(28.0-u[3]) - u[2]
+ du[3] = u[1]*u[2] - (8/3)*u[3]
+end
+u0 = [1.0;0.0;0.0]
+tspan = (0.0,100.0)
+prob = ODEProblem(lorenz,u0,tspan)
+sol = solve(prob,CVODE_Adams(),reltol=1e-12,abstol=1e-12)
+prob2 = ODEProblem(lorenz,sol[end],(100.0,0.0))
+sol = solve(prob,CVODE_Adams(),reltol=1e-12,abstol=1e-12)
+@show sol[end]-u0 #[-17.5445, -14.7706, 39.7985]
+```
+
+(here we once again use the CVODE C++ solvers from SUNDIALS since they are a close
+match to the SciPy integrators used in the neural ODE paper)
+
+This inaccuracy is the reason why the method from the neural ODE paper is not
+implemented in software suites, but it once again highlights a detail. Not
+all ODEs will have a large error due to this issue. And on ODEs where it's not
+a problem, this will be the most efficient way to do adjoint sensitivity
+analysis. Thus once again we arrive at the conclusion that one method is not
+enough.
+
+In DifferentialEquations.jl we are investigating many different methods for
+computing the derivatives of differential equations with respect to parameters.
+We have a [recent preprint](https://arxiv.org/abs/1812.01892) detailing
+some of these results. One of the things we have found is that direct use of
+automatic differentiation can be one of the most efficient and flexible methods.
+Julia's ForwardDiff.jl, Flux.jl, and ReverseDiff.jl can directly be applied to
+perform automatic differentiation on the native Julia differential equation

```

Comment:
It's maybe worth mentioning explicitly that this is something only Julia can do well because AD packages in C/C++/Fortran aren't suitable/available/etc. while reimplementing an ODE solver in terms of PyTorch or TensorFlow primitives would be much harder and end up much slower?

Reply:
Added a paragraph which makes this explicit.